tensorizer/serialization.py

```diff
@@ -4446,17 +4128,492 @@ def write_state_dict(self, state_dict: Union[Dict, List, Tuple]):
+            )
+
+    def _prepare_for_write_dtype(self, write_specs: Sequence[_WriteSpec]):
+        torch_dtype_to_numpy_dtype_cache: Dict[str, str] = {}
+
+        for w in write_specs:
+            tensor_dtype_str = str(w.tensor.dtype)
+            if _NumpyTensor._is_asymmetric(w.tensor.dtype):
+                # is opaque
+                w.dtype = (
+                    f"<V{w.tensor.element_size():d}"
+                    + OPAQUE_DTYPE_SEP
+                    + tensor_dtype_str
+                )
+                w.set_min_file_version_number(OPAQUE_TENSORIZER_VERSION)
+            else:
+                w.dtype = torch_dtype_to_numpy_dtype_cache.get(
+                    tensor_dtype_str, None
+                )
+                if w.dtype is None:
+                    torch_dtype_to_numpy_dtype_cache[tensor_dtype_str] = (
+                        w.dtype
+                    ) = (
+                        torch.tensor((), dtype=w.tensor.dtype, device="cpu")
+                        .numpy()
+                        .dtype.str
+                    )
+
+    @staticmethod
+    def _do_clone(write_spec, dependency: Optional[_Future]):
+        if dependency is not None:
+            dependency.result(_TIMEOUT)
+        write_spec.tensor = write_spec.tensor.clone().detach()
+        write_spec.numpy_tensor = _NumpyTensor.from_tensor(write_spec.tensor)
+
+    def _prepare_for_write_encryption(
+        self, write_specs: Sequence[_WriteSpec]
+    ) -> None:
+        assert self._encrypted and self._encryption is not None
+
+        # If any tensors are shared, so we need to clone all but one of them before encrypting
+        write_specs_by_addr: Dict[int, List[TensorSerializer._WriteSpec]] = (
+            collections.defaultdict(list)
+        )
+        for w in write_specs:
+            if w.tensor.device.type != "cpu":
+                continue
+            address = w.tensor.untyped_storage().data_ptr()
+            write_specs_by_addr[address].append(w)
+
+        for shared_write_specs in write_specs_by_addr.values():
+            if len(shared_write_specs) == 1:
+                continue
+
+            clone_dependencies = _FutureGroup(
+                [
+                    w.tensor_data_task
+                    for w in shared_write_specs
+                    if w.tensor_data_task is not None
+                ]
+            )
+
+            clone_tasks: List[_Future] = []
+            for w in shared_write_specs[1:]:
+                clone_tasks.append(
+                    self._computation_pool.submit(
+                        self._do_clone, w, clone_dependencies
+                    )
+                )
+                w.user_owns_tensor_data = False
+
+            shared_write_specs[0].tensor_data_task = _FutureGroup(
+                clone_tasks + clone_dependencies.futures
+            )
+            for w in shared_write_specs[1:]:
+                w.tensor_data_task = _FutureGroup(clone_tasks)
```

Comment:
I would choose the non-cloned tensor to be the last one, because its `tensor_data_task` will block the longest, so that would unblock some processing for a subsequent stage sooner, considering the access pattern of `write_specs` is consistently first-to-last.
```suggestion
            for w in shared_write_specs[:-1]:
                clone_tasks.append(
                    self._computation_pool.submit(
                        self._do_clone, w, clone_dependencies
                    )
                )
                w.user_owns_tensor_data = False

            shared_write_specs[-1].tensor_data_task = _FutureGroup(
                clone_tasks + clone_dependencies.futures
            )
            for w in shared_write_specs[:-1]:
                w.tensor_data_task = _FutureGroup(clone_tasks)
```